# Assignment 3

## Instructions - Read this first!

This is an individual homework assignment. This means that:

- You may discuss the problems in this assignment with other students in this course and your instructor/TA, but YOUR WORK MUST BE YOUR OWN.
- Do not show other students code or your own work on this assignment.
- You may consult external references, but not actively receive help from individuals not involved in this course.
- Cite all references outside of the course you used, including conversations with other students which were helpful. (This helps us give credit where it is due!). All references must use a commonly accepted reference format, for example, APA or IEEE (or another citation style of your choice).

If any of these rules seem ambiguous, please check with with your instructor for help interpreting them.

We suggest completing this assignment using the provided notebook. Each question should be answered using a SQL query (or combination or SQL queries) unless the text indicates that you may (or should) do something else. You may submit your queries embedded in Python, using SQLAlchemy or the MySQL Connector, or as plain text in Markdown.

## When you submit your work

Your submission will be graded manually. To ensure that everything goes smoothly, please follow these instructions to prepare your notebook for submission to the D2L Dropbox for Assignment 3:

- Please remove any print statments used to test your work (you can comment them out)
- Please provide your solutions where asked; please do not alter any other parts of this notebook.
- If you need to add cells to test your code please move them to the end of the notebook before submission- or you may included your commented out answers and tests in the cells provided

## Introduction

In this assignment, you will continue to practice and extend your SQL skills, and compare your work to MongoDB.

We will be using two datasets from Open Data Calgary. The Census by Community 2019 dataset shows results of the 2019 Civic Census. This is a separate process from that used by Statistics Canada to produce the national census, some of which results we explored in the previous assignment. 

This census data counts dwelling units in the city and categorized the population within each unit, aggregating results by community. The original dataset has been partially cleaned for you, removing some columns which are not relevant to the assignment. However, given that there are still 75 columns remaining, you may want to use SQLAlchemy or another bulk load tool to handle the creation of the database table for this dataset, before altering the table as needed to change data types or creating indexes.

The Community Services dataset lists a number of amenities in the city with their address and community. This table has not been altered from the version provided by Open Data Calgary.

## Data cleaning and import

First, import the two CSVs (Community_Services.csv, Census_by_Community_2019.csv) into your own database. You may use what is available to you on datasciencedb or datasciencedb2. You may also create indexes and define keys if appropriate for the column(s) of your choice.

In the section below, you have the option to discuss any data cleaning and wrangling steps performed during this process. This is not a requirement and will not be assessed directly for grading; however, this may help to clarify to your reader exactly what was done, to make your work below more understandable.

NOTES: All references will be attached to the bottom of this notebook.

In [1]:
import mysql.connector
from mysql.connector import errorcode

# uncomment the lines below and fill in any relevant details that need to be changed here, such as if you set up a different user or password

myconnection = mysql.connector.connect(user='khizer_kamran1', password='8VCW81ULC',host='datasciencedb2.ucalgary.ca', database='khizer_kamran1')
myconnection

In [2]:
import pandas as pd

cpl_locations = pd.read_csv("Community_Services.csv")
cpl_locations = cpl_locations.set_axis(['TYPE', 'NAME', 'ADDRESS', 'COMMCODE', 'POINT'], axis=1, inplace=False)
cpl_locations = cpl_locations.fillna(0)
cpl_locations

TYPE                                       NAME  \
0    Community Centre                  Rosemont Community Centre   
1          Attraction             WinSport's Canada Olympic Park   
2    Community Centre                   Mid-Sun Community Centre   
3          Attraction                               Arts Commons   
4    Community Centre  Willow Park / Mapleridge Community Centre   
..                ...                                        ...   
201  Community Centre     Forest Heights/ Fonda Community Centre   
202  Community Centre          Cambrian Heights Community Centre   
203  Community Centre               Banff Trail Community Centre   
204  Community Centre               Rosscarrock Community Centre   
205  Community Centre            Canyon Meadows Community Centre   

                     ADDRESS COMMCODE                            POINT  
0              2807 10 ST NW      CAP   POINT (-114.0860375 51.076753)  
1    88 Canada Olympic RD SW      COP  POINT (-114.2154069 51.0823997)  
2           50 Midpark RI SE      MID  POINT (-114.0566252 50.9154377)  
3                205 8 AV SE      DNC  POINT (-114.0600555 51.0453422)  
4           680 Acadia DR SE      MPL  POINT (-114.0433391 50.9598927)  
..                       ...      ...                              ...  
201        4909 Forego AV SE      FHT   POINT (-113.963017 51.0492163)  
202     600 Northmount DR NW      CAM   POINT (-114.087382 51.0868288)  
203            2115 20 AV NW      BNF  POINT (-114.1089218 51.0703032)  
204            4411 10 AV SW      RCK  POINT (-114.1507529 51.0427976)  
205     848 Cantabrian DR SW      CAN  POINT (-114.0788446 50.9462137)  

[206 rows x 5 columns]

Notes: I re-named the header names of my Community_Services table to make it easier to insert into SQL and I repkaced all rows containing NaN values with 0s; cleaning process.

In [3]:
import numpy as np
import pandas as pd

cpl_locations2 = pd.read_csv("Census_by_Community_2019.csv")
cpl_locations2 = cpl_locations2.set_axis(['CLASS', 'CODE', 'COMMCODE', 'NAME', 'SECTOR', 'SRG', 'COMMSTRUCTURE', 'CNSSYR', 'FOIPIND', 'RESCNT', 'DWELLCNT', 'PRSCHCHLD', 'OWNSHPCNT', 'PUBSCH', 'SEPSCH', 'PUBSEPSCH', 'OTHERSCH', 'UNKNWNSCH', 'SINGFAMLY', 'DUPLEX', 'MULTIPLEX', 'APARTMENT', 'TOWNHOUSE', 'MANUFHOME', 'CONVSTRUC', 'COMUNLHSE', 'RESCOMM', 'OTHERRES', 'NURSINGHM', 'OTHERINST', 'HOTELCNT', 'OTHERMISC', 'MALECNT', 'FEMALECNT', 'MALE04', 'MALE514', 'MALE1519', 'MALE2024', 'MALE2534', 'MALE3544', 'MALE4554', 'MALE5564', 'MALE6574', 'MALE7', 'FEM04', 'FEM514', 'FEM1519', 'FEM2024', 'FEM2534', 'FEM3544', 'FEM4554', 'FEM5564', 'FEM6574', 'FEM75', 'MF04', 'MF514', 'MF1519', 'MF2024', 'MF2534', 'MF3544', 'MF4554', 'MF5564', 'MF6574', 'MF75', 'OTHERCNT', 'OTHER04', 'OTHER514', 'OTHER1519', 'OTHER2024', 'OTHER2534', 'OTHER3544', 'OTHER4554', 'OTHER5564', 'OTHER6574', 'OTHER75'], axis=1, inplace=False)
cpl_locations2 = cpl_locations2.fillna(0)
cpl_locations2

CLASS  CODE COMMCODE           NAME     SECTOR         SRG  \
0          Residential     1      LEG         LEGACY      SOUTH  DEVELOPING   
1          Residential     1      HPK  HIGHLAND PARK     CENTRE   BUILT-OUT   
2          Residential     1      CNS    CORNERSTONE  NORTHEAST  DEVELOPING   
3          Residential     1      MON     MONTGOMERY  NORTHWEST   BUILT-OUT   
4          Residential     1      TEM         TEMPLE  NORTHEAST   BUILT-OUT   
..                 ...   ...      ...            ...        ...         ...   
301  Residual Sub Area     4      01H            01H       WEST           0   
302        Residential     1      HID  HIDDEN VALLEY      NORTH   BUILT-OUT   
303        Residential     1      RIV      RIVERBEND  SOUTHEAST   BUILT-OUT   
304        Residential     1      RID    RIDEAU PARK     CENTRE   BUILT-OUT   
305         Industrial     2      FRA       FRANKLIN  NORTHEAST           0   

    COMMSTRUCTURE  CNSSYR FOIPIND  RESCNT  ...  OTHER04  OTHER514  OTHER1519  \
0    BUILDING OUT    2019       0    6420  ...        0         0          0   
1           1950s    2019       0    3838  ...        0         0          0   
2           2000s    2019       0    2648  ...        0         0          0   
3           1950s    2019       0    4515  ...        0         0          0   
4     1960s/1970s    2019       0   10977  ...        0         0          0   
..            ...     ...     ...     ...  ...      ...       ...        ...   
301   UNDEVELOPED    2019       Y       0  ...        0         0          0   
302   1980s/1990s    2019       0   11566  ...        0         6          6   
303   1980s/1990s    2019       0    9244  ...        0         0          0   
304    INNER CITY    2019       0     594  ...        0         0          0   
305    EMPLOYMENT    2019       0       0  ...        0         0          0   

     OTHER2024  OTHER2534  OTHER3544  OTHER4554  OTHER5564  OTHER6574  OTHER75  
0            0          0          0          0          0          0        0  
1            0          0          0          0          0          0        0  
2            0          0          0          0          0          0        0  
3            0          0          0          0          0          0        0  
4            0          0          0          0          0          0        0  
..         ...        ...        ...        ...        ...        ...      ...  
301          0          0          0          0          0          0        0  
302          6          0          0          9          0          0        0  
303          0          0          0          0          0          0        0  
304          0          0          0          0          0          0        0  
305          0          0          0          0          0          0        0  

[306 rows x 75 columns]

Notes: I re-named the header names of my Community_Services table to make it easier to insert into SQL and I repkaced all rows containing NaN values with 0s; cleaning process.

In [4]:
# CREATE TABLE STATEMENT
create_statement = '''create table Community_Services (
    TYPE varchar(100),
    NAME varchar(100),
    ADDRESS varchar(100),
    COMMCODE varchar(10),
    POINT varchar(100)
    );'''

# now we'll create a cursor and run our create statement
create_cursor = myconnection.cursor()
try:
    create_cursor.execute(create_statement)
except mysql.connector.Error as err:
    if err.errno == errorcode.ER_TABLE_EXISTS_ERROR:
        print("Ooops! We already have that table")
    else:
        print(err.msg)
else:
    print("table created successfully!")

create_cursor.close()

Ooops! We already have that table


True

In [5]:
# CREATE TABLE STATEMENT
create_statement = '''create table Census_Community (
    CLASS varchar(100), 
    CODE int, 
    COMMCODE varchar(100), 
    NAME varchar(100), 
    SECTOR varchar(100), 
    SRG varchar(100), 
    COMMSTRUCTURE varchar(100), 
    CNSSYR int, 
    FOIPIND varchar(100), 
    RESCNT int, 
    DWELLCNT int, 
    PRSCHCHLD int, 
    OWNSHPCNT int, 
    PUBSCH int, 
    SEPSCH int, 
    PUBSEPSCH int, 
    OTHERSCH int, 
    UNKNWNSCH int, 
    SINGFAMLY int, 
    DUPLEX int, 
    MULTIPLEX int, 
    APARTMENT int, 
    TOWNHOUSE int, 
    MANUFHOME int, 
    CONVSTRUC int, 
    COMUNLHSE int, 
    RESCOMM int, 
    OTHERRES int, 
    NURSINGHM int, 
    OTHERINST int, 
    HOTELCNT int, 
    OTHERMISC int, 
    MALECNT int, 
    FEMALECNT int, 
    MALE04 int, 
    MALE514 int, 
    MALE1519 int, 
    MALE2024 int, 
    MALE2534 int, 
    MALE3544 int, 
    MALE4554 int, 
    MALE5564 int, 
    MALE6574 int, 
    MALE7 int, 
    FEM04 int, 
    FEM514 int, 
    FEM1519 int, 
    FEM2024 int, 
    FEM2534 int, 
    FEM3544 int, 
    FEM4554 int, 
    FEM5564 int, 
    FEM6574 int, 
    FEM75 int, 
    MF04 int, 
    MF514 int, 
    MF1519 int, 
    MF2024 int, 
    MF2534 int, 
    MF3544 int, 
    MF4554 int, 
    MF5564 int, 
    MF6574 int, 
    MF75 int, 
    OTHERCNT int, 
    OTHER04 int, 
    OTHER514 int, 
    OTHER1519 int, 
    OTHER2024 int, 
    OTHER2534 int, 
    OTHER3544 int, 
    OTHER4554 int, 
    OTHER5564 int, 
    OTHER6574 int, 
    OTHER75 int
    );'''

# now we'll create a cursor and run our create statement
create_cursor = myconnection.cursor()
try:
    create_cursor.execute(create_statement)
except mysql.connector.Error as err:
    if err.errno == errorcode.ER_TABLE_EXISTS_ERROR:
        print("Ooops! We already have that table")
    else:
        print(err.msg)
else:
    print("table created successfully!")

create_cursor.close()

Ooops! We already have that table


True

In [6]:
# insertCursor = myconnection.cursor()

# columnString = "`,`".join([str(currentColumn) for currentColumn in cpl_locations.columns.tolist()])
# #print (columnString)

# # inserting rows one by one from the DataFrame is sufficient for now
# for i, currentRow in cpl_locations.iterrows():
#     print (tuple(currentRow))
#     insertCommand = "INSERT INTO `Community_Services` (`" + columnString + "`) VALUES (" + "%s,"*(len(currentRow)-1) + "%s)"
#     print (insertCommand)
#     insertCursor.execute(insertCommand, tuple(currentRow))
    
# myconnection.commit()

# insertCursor.close()

In [7]:
# insertCursor = myconnection.cursor()

# columnString = "`,`".join([str(currentColumn) for currentColumn in cpl_locations2.columns.tolist()])
# #print (columnString)

# # inserting rows one by one from the DataFrame is sufficient for now
# for i, currentRow in cpl_locations2.iterrows():
#     print (tuple(currentRow))
#     insertCommand = "INSERT INTO `Census_Community` (`" + columnString + "`) VALUES (" + "%s,"*(len(currentRow)-1) + "%s)"
#     print (insertCommand)
#     insertCursor.execute(insertCommand, tuple(currentRow))
    
# myconnection.commit()

# insertCursor.close()

In [8]:
# mycursor = myconnection.cursor()

# sql = "DROP TABLE Community_Services"

# mycursor.execute(sql)

In [9]:
import sqlalchemy as sq
sq.__version__

engine = sq.create_engine('mysql+mysqlconnector://khizer_kamran1:8VCW81ULC@datasciencedb2.ucalgary.ca/khizer_kamran1')

## Part A: Warm-up Questions (10 marks)

Answer the questions below, including any queries you used where necessary. Not all questions will require a SQL query for a correct response. You may wish to use as a source the references which are already provided as part of this notebook.

First, let's look at the Census by Community Data.

1. How many communities are included as part of this census? **(1 mark)**

2. There are 75 columns in the version of the dataset provided to you. While not all of them are equally interesting, briefly describe the following columns. **(3 marks)**

For Question 2:

   - CLASS: Refers to the various sections within a metropolitan area that corresponds to what groups of people can be found there; residential will contain citizens, industrial will contain workers, major park will contain a variety of citizens and wildlife etc...

   - SRG (include in your answer all posslbe values): Whether or not a particular housing community has previously acquired tenants or not; in Built-Out communities the properties have been developed and tenants have lived within those communities previously where as for Developing the tenants may not have actually lived there as of yet as the properties within the communities are still being developed or finalized.

   - COMM_STUCTURE: Means the internal structure of an employment, area, town, city, neighbourhood or another residential/urban areas. It includes the population and housing, jobs and production, service and leisure time areas, along with transport routes and technical networks, their location and relationships.
   

3. What other columns, to your eye, look interesting? Provide an example of a guiding question which would take advantage of one of these columns. **(2 marks)**

In [10]:
#For Question 1:

query_table = pd.read_sql_query('SELECT COUNT(DISTINCT NAME) AS CountDistinctCommunities FROM Census_Community;', engine)
print(query_table)
print("The number of unique communities is 306.")

   CountDistinctCommunities
0                       306
The number of unique communities is 306.


For Question 3:

Potential Guiding Question (looking at the SECTOR column) - Which sector of the city (using the SECTOR column) appears to contain the most amount of Built-Out communities?

Next, let's look at the Community Services table. 

4. How many services in total are there? **(1 mark)**


5. List all possible types of community services listed in the table. How many are missing a community code? **(1 mark)**


6. Which column is the most useful as a key to be used in a join? **(1 mark)**


7. Which community has the highest number of available community services? Include the names of any community with this number of services. **(1 mark)**

In [11]:
#For Question 4:

query_table = pd.read_sql_query('SELECT DISTINCT TYPE AS CountDistinctServices FROM Community_Services;', engine)
print(query_table)
print('\n')
print("The number of unique services is 8.")

  CountDistinctServices
0      Community Centre
1            Attraction
2          Visitor Info
3                 Court
4               Library
5              Hospital
6            PHS Clinic
7        Social Dev Ctr


The number of unique services is 8.


In [12]:
#For Question 5:

query_table = pd.read_sql_query('SELECT DISTINCT TYPE AS DistinctServices FROM Community_Services WHERE COMMCODE = "0";', engine)
print(query_table)
print("Only 1 out of the 8 are missing a community code; AttractionThe number of unique services is 8.")

  DistinctServices
0       Attraction
Only 1 out of the 8 are missing a community code; AttractionThe number of unique services is 8.


In [13]:
#For Question 5:

query_table = pd.read_sql_query('SELECT DISTINCT TYPE AS DistinctService FROM Community_Services WHERE COMMCODE = "0";', engine)
print(query_table)
print("Only 1 out of 8 of the services provided are missing a community code (Attraction).")

  DistinctService
0      Attraction
Only 1 out of 8 of the services provided are missing a community code (Attraction).


For Question 6:

The COMMCODE column should be used as the key for join as this is present within both of the .csv files.

In [14]:
#For Question 7:
query_table = pd.read_sql_query('SELECT NAME, COMMCODE, TYPE AS DistinctType FROM Community_Services WHERE COMMCODE = "BLN";', engine)
query_table

NAME COMMCODE      DistinctType
0       Tourism Calgary Central Office      BLN      Visitor Info
1                Scotiabank Saddledome      BLN        Attraction
2  Beltline Neighbourhoods Association      BLN  Community Centre
3      Sheldon M. Chumir Health Centre      BLN        PHS Clinic
4                Memorial Park Library      BLN           Library
5                Calgary Stampede Park      BLN        Attraction

In [15]:
print("BLN Community Code has the most distinct services: 5")

BLN Community Code has the most distinct services: 5


## Part B: SQL with multiple tables (13 marks)


What is the population of seniors (male and female, or other, aged 65 and over) in the community or communities with the most community services? **(2 marks)**

What is the population of children (male and female, or other, aged 0 to 14), in the community or communities with the most community services?  **(2 marks)**

What is the total population of seniors (male and female, or other, aged 65 and over) in all communities that contain nursing homes?  **(2 marks)**

What is the total number of residents, children (using the definition above) and seniors (using the definitions above) in communities which do not have any community services?  **(2 marks)**

Generate a list of the ten such communities with the greatest number of residents.  **(2 marks)**

Prepare a list of community services as follows: **(3 marks)**
- Each community service should list, for the community which they are in, the total number of residents, the total number of preschool children, and the number of dwellings.
- Community services with no community listed should use a meaningful placeholder in the result.
- Communities with no community services should be included in the results, with a meaningful placeholder where data for community services would ordinarily be included.

Note: I used an RIGHT join as I wanted to keep all records based upon the Census_Community table that would help me the following questions.

In [16]:
query_table = pd.read_sql_query('SELECT * FROM Community_Services RIGHT JOIN Census_Community ON Community_Services.COMMCODE = Census_Community.COMMCODE;', engine)
query_table

TYPE                                       NAME  \
0    Community Centre                  Rosemont Community Centre   
1          Attraction             WinSport's Canada Olympic Park   
2    Community Centre                   Mid-Sun Community Centre   
3          Attraction                               Arts Commons   
4    Community Centre  Willow Park / Mapleridge Community Centre   
..                ...                                        ...   
370              None                                       None   
371              None                                       None   
372              None                                       None   
373              None                                       None   
374              None                                       None   

                     ADDRESS COMMCODE                            POINT  \
0              2807 10 ST NW      CAP   POINT (-114.0860375 51.076753)   
1    88 Canada Olympic RD SW      COP  POINT (-114.2154069 51.0823997)   
2           50 Midpark RI SE      MID  POINT (-114.0566252 50.9154377)   
3                205 8 AV SE      DNC  POINT (-114.0600555 51.0453422)   
4           680 Acadia DR SE      MPL  POINT (-114.0433391 50.9598927)   
..                       ...      ...                              ...   
370                     None     None                             None   
371                     None     None                             None   
372                     None     None                             None   
373                     None     None                             None   
374                     None     None                             None   

                 CLASS  CODE COMMCODE                      NAME     SECTOR  \
0          Residential     1      CAP              CAPITOL HILL     CENTRE   
1           Major Park     3      COP       CANADA OLYMPIC PARK       WEST   
2          Residential     1      MID                 MIDNAPORE      SOUTH   
3          Residential     1      DNC  DOWNTOWN COMMERCIAL CORE     CENTRE   
4          Residential     1      MPL               MAPLE RIDGE      SOUTH   
..                 ...   ...      ...                       ...        ...   
370        Residential     1      SPH           SPRINGBANK HILL       WEST   
371  Residual Sub Area     4      01H                       01H       WEST   
372        Residential     1      HID             HIDDEN VALLEY      NORTH   
373        Residential     1      RID               RIDEAU PARK     CENTRE   
374         Industrial     2      FRA                  FRANKLIN  NORTHEAST   

     ... OTHER04 OTHER514  OTHER1519 OTHER2024  OTHER2534  OTHER3544  \
0    ...       0        0          0         0          0          0   
1    ...       0        0          0         0          0          0   
2    ...       0        0          0         0          0          0   
3    ...       9       12          0        19         42         33   
4    ...       0        0          0         0          0          0   
..   ...     ...      ...        ...       ...        ...        ...   
370  ...       0        0          0         0          0          0   
371  ...       0        0          0         0          0          0   
372  ...       0        6          6         6          0          0   
373  ...       0        0          0         0          0          0   
374  ...       0        0          0         0          0          0   

     OTHER4554  OTHER5564  OTHER6574  OTHER75  
0            0          0          0        0  
1            0          0          0        0  
2            0          0          0        0  
3            0          0          0        0  
4            0          0          0        0  
..         ...        ...        ...      ...  
370          0          0          0        0  
371          0          0          0        0  
372          9          0          0        0  
373          0          0    

In [17]:
#For Question 1:
query_table = pd.read_sql_query('SELECT *, (Census_Community.MF6574 + Census_Community.MF75 + Census_Community.OTHER6574 + Census_Community.OTHER75) AS TotalPop FROM Community_Services RIGHT JOIN Census_Community ON Community_Services.COMMCODE = Census_Community.COMMCODE WHERE Census_Community.COMMCODE = "BLN";', engine)
query_table

TYPE                                 NAME  \
0      Visitor Info       Tourism Calgary Central Office   
1        Attraction                Scotiabank Saddledome   
2  Community Centre  Beltline Neighbourhoods Association   
3        PHS Clinic      Sheldon M. Chumir Health Centre   
4           Library                Memorial Park Library   
5        Attraction                Calgary Stampede Park   

                ADDRESS COMMCODE                            POINT  \
0      200 238 11 AV SW      BLN  POINT (-114.0675212 51.0427492)   
1  555 Saddledome RI SE      BLN  POINT (-114.0519704 51.0374013)   
2          102 17 AV SW      BLN   POINT (-114.063836 51.0379958)   
3        1213 - 4 ST SW      BLN  POINT (-114.0721296 51.0411634)   
4          1221 2 ST SW      BLN   POINT (-114.0690596 51.041109)   
5    1410 Olympic WY SE      BLN  POINT (-114.0536299 51.0391375)   

         CLASS  CODE COMMCODE      NAME  SECTOR  ... OTHER514 OTHER1519  \
0  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   
1  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   
2  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   
3  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   
4  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   
5  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   

   OTHER2024 OTHER2534  OTHER3544  OTHER4554  OTHER5564  OTHER6574  OTHER75  \
0         10       631         85         19          7          0        0   
1         10       631         85         19          7          0        0   
2         10       631         85         19          7          0        0   
3         10       631         85         19          7          0        0   
4         10       631         85         19          7          0        0   
5         10       631         85         19          7          0        0   

   TotalPop  
0      1605  
1      1605  
2      1605  
3      1605  
4      1605  
5      1605  

[6 rows x 81 columns]

The Total Population of Senior Citizens (65 years old and above) within the BLN community is: 1,605

In [18]:
#For Question 2:

query_table = pd.read_sql_query('SELECT *, (Census_Community.MF04 + Census_Community.MF514 + Census_Community.OTHER04 + Census_Community.OTHER514) AS TotalPop FROM Community_Services RIGHT JOIN Census_Community ON Community_Services.COMMCODE = Census_Community.COMMCODE WHERE Census_Community.COMMCODE = "BLN";', engine)
query_table

TYPE                                 NAME  \
0      Visitor Info       Tourism Calgary Central Office   
1        Attraction                Scotiabank Saddledome   
2  Community Centre  Beltline Neighbourhoods Association   
3        PHS Clinic      Sheldon M. Chumir Health Centre   
4           Library                Memorial Park Library   
5        Attraction                Calgary Stampede Park   

                ADDRESS COMMCODE                            POINT  \
0      200 238 11 AV SW      BLN  POINT (-114.0675212 51.0427492)   
1  555 Saddledome RI SE      BLN  POINT (-114.0519704 51.0374013)   
2          102 17 AV SW      BLN   POINT (-114.063836 51.0379958)   
3        1213 - 4 ST SW      BLN  POINT (-114.0721296 51.0411634)   
4          1221 2 ST SW      BLN   POINT (-114.0690596 51.041109)   
5    1410 Olympic WY SE      BLN  POINT (-114.0536299 51.0391375)   

         CLASS  CODE COMMCODE      NAME  SECTOR  ... OTHER514 OTHER1519  \
0  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   
1  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   
2  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   
3  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   
4  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   
5  Residential     1      BLN  BELTLINE  CENTRE  ...       14         0   

   OTHER2024 OTHER2534  OTHER3544  OTHER4554  OTHER5564  OTHER6574  OTHER75  \
0         10       631         85         19          7          0        0   
1         10       631         85         19          7          0        0   
2         10       631         85         19          7          0        0   
3         10       631         85         19          7          0        0   
4         10       631         85         19          7          0        0   
5         10       631         85         19          7          0        0   

   TotalPop  
0      1222  
1      1222  
2      1222  
3      1222  
4      1222  
5      1222  

[6 rows x 81 columns]

The Total Population of Children (0 - 14) within the BLN community is: 1,222

In [19]:
#For Question 3:

query_table = pd.read_sql_query('SELECT DISTINCT Census_Community.COMMCODE, SUM(Census_Community.MF6574 + Census_Community.MF75 + Census_Community.OTHER6574 + Census_Community.OTHER75) AS TotalPop FROM Community_Services RIGHT JOIN Census_Community ON Community_Services.COMMCODE = Census_Community.COMMCODE WHERE Census_Community.NURSINGHM > 0;', engine)
query_table

COMMCODE  TotalPop
0      CAP   91951.0

The Total Population of Seniors (65+) Amongst All Communities With 1 or More Nursing Homes is: 91,951

In [20]:
#For Question 4:

query_table = pd.read_sql_query('SELECT *, SUM(Census_Community.MF6574 + Census_Community.MF75 + Census_Community.OTHER6574 + Census_Community.OTHER75 + Census_Community.MF04 + Census_Community.MF514 + Census_Community.OTHER04 + Census_Community.OTHER514) AS TotalPop FROM Community_Services RIGHT JOIN Census_Community ON Community_Services.COMMCODE = Census_Community.COMMCODE WHERE Community_Services.TYPE IS NULL;', engine)
query_table

TYPE  NAME ADDRESS COMMCODE POINT        CLASS  CODE COMMCODE    NAME  \
0  None  None    None     None  None  Residential     1      LEG  LEGACY   

  SECTOR  ... OTHER514 OTHER1519  OTHER2024 OTHER2534  OTHER3544  OTHER4554  \
0  SOUTH  ...        0         0          0         0          0          0   

   OTHER5564  OTHER6574  OTHER75  TotalPop  
0          0          0        0  156162.0  

[1 rows x 81 columns]

The Total Population of Children and Seniors (0 - 14 and 65 - 75 respectively) within the BLN community is: 156,162

In [21]:
#For Question 5:

query_table = pd.read_sql_query('SELECT DISTINCT Census_Community.COMMCODE, (Census_Community.MF04 + Census_Community.MF514 + Census_Community.MF1519 + Census_Community.MF2024 + Census_Community.MF2534 + Census_Community.MF3544 + Census_Community.MF4554 + Census_Community.MF5564 + Census_Community.MF6574 + Census_Community.MF75 + Census_Community.Other04 + Census_Community.Other514 + Census_Community.Other1519 + Census_Community.Other2024 + Census_Community.Other2534 + Census_Community.Other3544 + Census_Community.Other4554 + Census_Community.Other5564 + Census_Community.Other6574 + Census_Community.Other75) AS TotalPop FROM Community_Services RIGHT JOIN Census_Community ON Community_Services.COMMCODE = Census_Community.COMMCODE ORDER BY TotalPop DESC;', engine)
query_table

COMMCODE  TotalPop
0        PAN     25918
1        BLN     25874
2        SAD     22325
3        EVE     21527
4        TUS     19878
..       ...       ...
301      ST2         0
302      12I         0
303      06C         0
304      BVD         0
305      SGL         0

[306 rows x 2 columns]

List of 10 Communities With the Largest Populations:

1. Panorama Hills - PAN
2. Beltline - BLN
3. Saddle Ridge - SAD
4. Evergreen - EVE
5. Tuscany - TUS
6. Cranston - CRA
7. Taradle - TAR
8. Mckenzie Towne - MCT
9. Coventry Hills - COV
10. Evanston - EVN

In [22]:
#For Question 6:

query_table = pd.read_sql_query('SELECT TYPE, ADDRESS,Census_Community.COMMCODE, Census_Community.NAME, DWELLCNT, PRSCHCHLD, (Census_Community.MF04 + Census_Community.MF514 + Census_Community.MF1519 + Census_Community.MF2024 + Census_Community.MF2534 + Census_Community.MF3544 + Census_Community.MF4554 + Census_Community.MF5564 + Census_Community.MF6574 + Census_Community.MF75 + Census_Community.Other04 + Census_Community.Other514 + Census_Community.Other1519 + Census_Community.Other2024 + Census_Community.Other2534 + Census_Community.Other3544 + Census_Community.Other4554 + Census_Community.Other5564 + Census_Community.Other6574 + Census_Community.Other75) AS TotalPop FROM Community_Services RIGHT JOIN Census_Community ON Community_Services.COMMCODE = Census_Community.COMMCODE ORDER BY Census_Community.NAME ASC;', engine)
df = pd.DataFrame(query_table)
df = df.fillna(value="No Service/Community Information Available")
df

TYPE  \
0    No Service/Community Information Available   
1    No Service/Community Information Available   
2    No Service/Community Information Available   
3    No Service/Community Information Available   
4    No Service/Community Information Available   
..                                          ...   
370                            Community Centre   
371  No Service/Community Information Available   
372  No Service/Community Information Available   
373                            Community Centre   
374  No Service/Community Information Available   

                                        ADDRESS COMMCODE  \
0    No Service/Community Information Available      01B   
1    No Service/Community Information Available      01C   
2    No Service/Community Information Available      01F   
3    No Service/Community Information Available      01H   
4    No Service/Community Information Available      01I   
..                                          ...      ...   
370                                520 27 AV NE      WIN   
371  No Service/Community Information Available      WWO   
372  No Service/Community Information Available      WBN   
373                         1991 Woodview DR SW      WOO   
374  No Service/Community Information Available      YKV   

                          NAME  DWELLCNT  PRSCHCHLD  TotalPop  
0                          01B         0          0         0  
1                          01C         0          0         0  
2                          01F         0          0         0  
3                          01H         0          0         0  
4                          01I         0          0         0  
..                         ...       ...        ...       ...  
370  WINSTON HEIGHTS/MOUNTVIEW      1803        247      3599  
371                WOLF WILLOW        21          0         0  
372                   WOODBINE      3372        557      8860  
373                  WOODLANDS      2411        438      5987  
374                  YORKVILLE       118          0         0  

[375 rows x 7 columns]

Note: I replaced the none values for service type and community address with "No Service/Community Information Available" as I felt this best represented how to indicate why there isn't information gathered for those fields.

## Part C: Evaluating your results  (7 marks)


**Question 1 (3 marks)**

The queries in Part B suggest that there are parts of Calgary which may be over-served by the numbers of community services available, and there are parts of Calgary which may be underserved by kinds of community services which are not available.

Before we continue with this line of inquiry, discuss two factors which may not be captured by either dataset.



Nominate one datset already available from Open Data Calgary which addresses one of your two factors. Remember to reference this using a well-accepted reference format.

For Question 1:

The two factors that the dataset does not consider are community income levels per community demographic and the amounts to which provincial government spending is allocated towards each of YYC's sectors and communities. Community income levels will determine the number of residences (AKA dwelling counts - usually the more money a community generates on a per-household basis the more expensive the community residential areas are and hence less citizens reside there due to higher costs of living). Community allocation of provincial government spending indicates the level and development of social services that are provided to various residential areas within YYC (if certain communities have a need for more clinics or more community centres - for example - these can be budgeted for and appropriately resolved).

Dataset Chosen: Calgary Equity Index Matrix (References are found at the bottom).

This dataset evaluates the standard of living and wellbeing of Calgarians based on both social as well as economic factors in which provides insights on the socio-economic positioning of various demographics and communities within the City of Calgary. This particular dataset can effectively address the first factor of community income levels and when paired with our previous queried datasets can help to provide more insights on trends found within social services found on a residence-by-residence basis.


**Question 2 (4 marks)**

Imagine that we have stored data relevant to each community in a separate JSON object, each of which has been added to a document in the same collection in MongoDB. Community Services have been added as list of JSON attributes to the relevant community.

For example, we might have
```
{
  Community:"Community Name", 
  ...
  Community_Services:[
      {
          "Community Service": "Community Service Name"
          
      },
      ...
      ]
  }
```
Pick one of the queries from Part B to discuss. Do you think it is more difficult to retrieve the information requested for this query from the pair of relational database tables provided to you, or from a MongoDB collection set up as described? Explain why or why not. 

From my interpretations of what has been instructed to me thus far and my experiences with both JSON and SQL, SQL is known for being a higher performance programming language that can allow for easy storage, collection, and parsing of data. NoSQL/MySQL (in Python) have allowed us to quickly read in tables from .CSV files (and other file format types) and from there we are able to easily apply 1 - 2 line queries to retrieve the data we would like. In JSON files the introductory steps of firstly setting up the correct formatting for the data structure in it of itself is quite time consuming and overall inefficient. Coupling this with having to read the data multi-dimensionally (or through multi-level indexing) allows us to parse the data in a less efficient manner. When looking at Question 5 from Part B querying with multiple conditions in 1 - 2 lines allows us to retrieve the data in an intuitive and simple manner; when looking at summation for this question it is built into the query itself as well. When parsing data with multiple conditions in a .JSON file structure we would have to set up the .JSON file code with nested for-loop and if-elif-else statements to help reach even one part of our answer. For example, Question 5 asks us to create a list of 10 communities with the highest residency counts; in SQL this is a matter of simply selecting the columns we would like to include, include a summation expression, and then include our join tables as well as where clauses. Whereas for .JSON, we would have to first start by indexing based upon the community names as a key then individually retrieve through more looping and parsing the information related to specific community services, and then collect their individual counts (which would be the value in the key:value pair), from there we would have to add the values manually as summation in this case would not create a separate key:value pair and finally we would have to ensure from the very beginning that both .CSV files were incorporated into the .JSON file structure together to replicate a branching or multi-dimensional-esque structure that will help with parsing with the usage of key:value pairs. From this comparison and example it is evident that SQL is much easier to retrieve information from and .JSON is much more difficult respectively.

## Part D: Reflection (5 marks)


In 100 to 250 words, identify a concept you have found difficult or confusing from this assignment. Reflect on how your previous learning or experience helped you to understand this concept. Provide your reflection using markdown in the cell below.



The main concept from this assignment that I particularly found challenging were two: understanding the joining of two tables and advanced filtering with SQL queries. With regards to the first challenge I used my critical-thinking skills to help decipher what table rows needed to remain after the join and by identifying patterns in the files I was able to determine what each row within the tables meant (explained in some detail earlier in this assignment). The usage of pattern identification had helped me to get a better understanding of the dataset overall and made my future querying processes much more effective as well as efficient. With regards to the completion of filtering SQL my first approach was to attempt each query through a step-by-step approach, where I would implement one component of what I wanted SQL to derive from the data and increment my code to include specific information through the addition of more filters. By incrementing my queries in such a way I was able to add the various pieces of the SQL filter sequentially and at the same time greatly reduced the time needed to complete my queries. My main approach within this second case was to firstly visualize what information I wanted to have SQL dissect for me; I would write out in simple language on a piece of paper what I wanted the SQL query to return to me (whilst also using the sequential methodology aforementioned). From there I would work to specify what information was necessary and non-necessary to help reduce the amount of lines (decreasing inefficiencies with my code in the process) to help optimize my queries. I would cross reference my answers with the Excel filtering tool to help ensure that I had received the values I was directly looking for. For the most part with regards to this assignment an adjustment in thinking and methodologies was required as I had already used SQL in past university coursework, it was merely how to adapt my previous experiences with the programming languages into these new tasks that I had aimed to resolve.

In [23]:
# Use this cell to include some code to dispose of your SQLAlchemy engine object
engine.dispose()
myconnection.close()

## References

Both datasets used in this Assignment are licensed under the Open Government License - City of Calgary.

Calgary Equity Index Matrix [online], 2022. Open Calgary. Available from: https://data.calgary.ca/Help-and-Information/Calgary-Equity-Index-Matrix/xeek-u7v8b [Accessed 4 Dec 2022].

Census by Community 2019 [online], 2022. Open Calgary. Available from: https://data.calgary.ca/Demographics/Census-by-Community-2019/rkfr-buzb [Accessed 26 Nov 2022].

Community Services [online], 2022. Open Calgary. Available from: https://data.calgary.ca/Services-and-Amenities/Community-Services/x34e-bcjz [Accessed 26 Nov 2022].

Living Enviornment and Planning. [online], 2022. YMParisto. Available from: https://www.ymparisto.fi/en-us/living_environment_and_planning/Community_structure [Accessed 4 Dec 2022].

University of Calgary (Faculty of Science - Computer Science) [online], 2022. Getting started with SQLAlchemy and pandas. Available from https://d2l.ucalgary.ca/d2l/le/content/472036/viewContent/5581075/View [Accessed 4 Dec 2022].
 
University of Calgary (Faculty of Science - Computer Science) [online], 2022. Getting started with Python SQL Connectors. Available from https://d2l.ucalgary.ca/d2l/le/content/472036/viewContent/5581071/View [Accessed 4 Dec 2022].
